core/src/transport/circuit_breaking.rs

```diff
@@ -0,0 +1,383 @@
+/// `Transport`. The delay for `addr` increases if dialing it fails repeatedly.
+///
+/// Open circuits are closed automatically after some time. In addition, an open
+/// circuit for an address is closed after successfully dialing that address.
+#[derive(Clone)]
+pub struct CircuitBreaking<TInner> {
+    /// The underlying transport.
+    inner: TInner,
+    state: Arc<CircuitBreakerState>,
+}
+
+/// Represents open circuits.
+struct CircuitBreakerState {
+    addresses: Mutex<FnvHashMap<Multiaddr, AddrInfo>>,
+}
+
+#[derive(Debug)]
+struct AddrInfo {
+    /// The delay to add before calling [`Transport::dial`].
+    wait_duration: Duration,
+    /// After `expiration` the circuit can be closed.
+    expiration: Delay,
+}
+
+impl CircuitBreakerState {
+    fn new() -> Self {
+        CircuitBreakerState {
+            addresses: Mutex::new(Default::default()),
+        }
+    }
+
+    /// Returns `None` if circuit of `addr` is closed.
+    fn get_wait_duration(&self, addr: &Multiaddr) -> Option<Duration> {
+        self.addresses
+            .lock()
+            .get(addr)
+            .map(|addr_info| addr_info.wait_duration)
+    }
+
+    fn open_or_extend_circuit(&self, addr: Multiaddr) {
+        match self.addresses.lock().entry(addr) {
+            Entry::Vacant(entry) => {
+                entry.insert(AddrInfo {
+                    wait_duration: INITIAL_TIMEOUT,
+                    expiration: Delay::new(INITIAL_TIMEOUT * 2),
+                });
+            }
+
+            Entry::Occupied(entry) => {
+                let addr_info = entry.into_mut();
+                addr_info.wait_duration = cmp::min(addr_info.wait_duration * 2, MAX_TIMEOUT);
+                addr_info.expiration = Delay::new(addr_info.wait_duration * 2);
+            }
+        }
+    }
+
+    /// Closes the circuit of `addr` even if the circuit has not yet expired.
+    fn close_circuit(&self, addr: &Multiaddr) {
+        self.addresses.lock().remove(addr);
+    }
+
+    fn close_expired_circuits(&self) {
+        let mut addresses = self.addresses.lock();
+        addresses.retain(move |_, a| (&mut a.expiration).now_or_never().is_none());
+    }
+}
+
+impl<TInner> CircuitBreaking<TInner> {
+    pub fn new(inner: TInner) -> Self {
+        CircuitBreaking {
+            inner,
+            state: Arc::new(CircuitBreakerState::new()),
+        }
+    }
+}
+
+impl<TInner> Transport for CircuitBreaking<TInner>
+where
+    TInner: Transport,
+    TInner::Error: 'static,
+{
+    type Output = TInner::Output;
+    type Error = TInner::Error;
+    type Listener = TInner::Listener;
+    type ListenerUpgrade = TInner::ListenerUpgrade;
+    type Dial = CircuitBreakingDial<TInner>;
+
+    fn listen_on(self, addr: Multiaddr) -> Result<Self::Listener, TransportError<Self::Error>> {
+        self.inner.listen_on(addr)
+    }
+
+    fn dial(self, addr: Multiaddr) -> Result<Self::Dial, TransportError<Self::Error>> {
```

Comment:
`addr` might include a `PeerId`.

In large public networks (e.g. IPFS) many nodes advertise LAN local addresses. In most cases nodes don't share the same LAN and thus dialing these addresses fails.

Ignoring the `PeerId` in an address in `CircuitBreakerState` would prevent dialing say `192.168.2.108` multiple times in a row.

What do you think?

Reply:
Good point! Done in f4c6f67.